firedrake/preconditioners/pmg.py

```diff
@@ -274,55 +303,250 @@ def __init__(self, Vf, Vc, Vf_bcs, Vc_bcs):
+        }else{
+           TA1 = notr;
+           TA2 = tran;
+        }
+        TA3 = TA2;
+
+        m = mx;  k = nx;  n = ny*nz*nel;
+        lda = (tflag>0)? nx : mx;
+        dgemm_(&TA1, &notr, &m,&n,&k, &one, A1,&lda, x,&k, &zero, y,&m);
+
+        p = 0;  s = 0;
+        m = mx;  k = ny;  n = my;
+        lda = (tflag>0)? ny : my;
+        for(int i=0; i<nz*nel; i++){
+           dgemm_(&notr, &TA2, &m,&n,&k, &one, y+p,&m, A2,&lda, &zero, x+s,&m);
+           p += m*k;
+           s += m*n;
+        }
+
+        p = 0;  s = 0;
+        m = mx*my;  k = nz;  n = mz;
+        lda = (tflag>0)? nz : mz;
+        for(int i=0; i<nel; i++){
+           dgemm_(&notr, &TA3, &m,&n,&k, &one, x+p,&m, A3,&lda, &zero, y+s,&m);
+           p += m*k;
+           s += m*n;
+        }
+        return;
+        }
+        """
 
+        # UFL elements order the component DoFs related to the same node contiguously.
+        # We transpose before and after the multiplcation times J to have each component
+        # stored contiguously as a scalar field, thus reducing the number of dgemm calls.
 
-void restriction(double *restrict Rc, const double *restrict Rf, const double *restrict w)
-{{
-    double Afc[{dimf}*{dimc}] = {{0}};
-    expression_kernel(Afc);
-    for (int32_t i = 0; i < {dimf}; i++)
-       for (int32_t j = 0; j < {dimc}; j++)
-           Rc[j] += Afc[i*{dimc} + j] * Rf[i] / w[i];
-}}
-"""
+        prolong_code = f"""
+        {kronmxv_code}
 
-        self.restrict_kernel = op2.Kernel(self.restrict_code, "restriction")
-        self.mesh = Vf.mesh()
+        void prolongation(double *restrict y, const double *restrict x){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {nxyz}*i] = x[i + {nscal}*j];
+
+            kronmxv(0, {mx},{my},{mz}, {nx},{ny},{nz}, {nscal}, JX,JX,{JZ}, t0,t1);
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                   y[i + {nscal}*j] = t1[j + {mxyz}*i];
+            return;
+        }}
+        """
+
+        restrict_code = f"""
+        {kronmxv_code}
 
+        void restriction(double *restrict y, const double *restrict x,
+        const double *restrict w){{
+            double JX[{mx}*{nx}] = {{ {JX} }};
+            double t0[{lwork}], t1[{lwork}];
+            double one=1.0E0;
+
+            for(int j=0; j<{mxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    t0[j + {mxyz}*i] = x[i + {nscal}*j] * w[i + {nscal}*j];
+
+            kronmxv(1, {nx},{ny},{nz}, {mx},{my},{mz}, {nscal}, JX,JX,{JZ}, t0,t1);
+
+            for(int j=0; j<{nxyz}; j++)
+                for(int i=0; i<{nscal}; i++)
+                    y[i + {nscal}*j] += t1[j + {nxyz}*i];
+            return;
+        }}
+        """
+
+        op2.configuration["ldflags"] = "-lblas"
```

Comment:
Don't do this, instead add extra `ldargs` to the `Kernel`s you make.

Reply:
can you give an example?

Reply:
`op2.Kernel("void foo(...)", "foo", ldargs=["-lblas"])`